### PR TITLE
add Cancel Interaction RPC

### PR DESCRIPTION
**Pull request by iCollin**
Implements/Fixes [#0184](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0184-cancel-interaction.md)

This PR is **ready** for review.

### Testing Plan
ATF Test Scripts

### Summary
Add the ability to allow interaction alerts to be canceled.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: ffw/UIRPC.js [hunk 1]
```diff
@@ -287,6 +287,33 @@ FFW.UI = FFW.RPCObserver.create(
             }
             break;
           }
+          case 'UI.CancelInteraction':
+          {
+            var targetID = request.params.cancelID;
+            var typeID = request.params.functionID;
+            
+            if (typeID === 10 && SDL.InteractionChoicesView.active
+               && (targetID === -1 || targetID === SDL.InteractionChoicesView.cancelID)) {
```
Comment (jacobkeeler) [inline ffw/UIRPC.js]: Do not use `-1` to indicate an omitted `cancelID`

Comment (iCollin) [reply ffw/UIRPC.js]: fixed in https://github.com/smartdevicelink/sdl_hmi/pull/203/commits/a68f82378c9c1c7035ba582649a932be7d0e2d4c

File: ffw/UIRPC.js [hunk 2]
```diff
@@ -287,6 +287,33 @@ FFW.UI = FFW.RPCObserver.create(
...
+              SDL.InteractionChoicesView.deactivate('ABORTED');
+            } else if (typeID === 12 && SDL.AlertPopUp.active
+               && (targetID === undefined || targetID === SDL.AlertPopUp.cancelID)) {
+              SDL.AlertPopUp.deactivate();
+            } else if (typeID === 25 && SDL.ScrollableMessage.active
+               && (targetID === undefined || targetID === SDL.ScrollableMessage.cancelID)) {
+              SDL.ScrollableMessage.deactivate(true);
+            } else if (typeID === 26 && SDL.SliderView.active
+               && (targetID === undefined || targetID === SDL.SliderView.cancelID)) {
+              SDL.SliderView.deactivate(true);
```
Comment (jacobkeeler) [inline ffw/UIRPC.js]:
```suggestion
              SDL.SliderView.deactivate(false);
```
The current value results in a `TIMED_OUT` result code.

File: ffw/UIRPC.js [hunk 3]
```diff
@@ -287,6 +287,33 @@ FFW.UI = FFW.RPCObserver.create(
...
+            var typeID = request.params.functionID;
+
+            if (typeID === 10 && SDL.InteractionChoicesView.active
+               && (targetID === undefined || targetID === SDL.InteractionChoicesView.cancelID)) {
+              SDL.InteractionChoicesView.deactivate('ABORTED');
+            } else if (typeID === 12 && SDL.AlertPopUp.active
+               && (targetID === undefined || targetID === SDL.AlertPopUp.cancelID)) {
+              SDL.AlertPopUp.deactivate();
+            } else if (typeID === 25 && SDL.ScrollableMessage.active
+               && (targetID === undefined || targetID === SDL.ScrollableMessage.cancelID)) {
```
Comment (jacobkeeler) [inline ffw/UIRPC.js]: Don't know why, but when I try to use a cancelID for ScrollableMessage, I get an `IGNORED` response.

Comment (iCollin) [reply ffw/UIRPC.js]: fixed in core commit https://github.com/smartdevicelink/sdl_core/pull/2963/commits/6774f3475722d88f6c552823c07ed36b8ac8ea62

File: ffw/UIRPC.js [hunk 4]
```diff
@@ -287,6 +287,33 @@ FFW.UI = FFW.RPCObserver.create(
...
+              SDL.ScrollableMessage.deactivate(true);
+            } else if (typeID === 26 && SDL.SliderView.active
+               && (targetID === undefined || targetID === SDL.SliderView.cancelID)) {
+              SDL.SliderView.deactivate(false);
```
Comment (jacobkeeler) [inline ffw/UIRPC.js]:
```suggestion
              SDL.SliderView.deactivate();
```
Apparently false indicates a `SUCCESS` result, no idea why. You apparently need to omit to parameter to send `ABORTED`.

File: ffw/UIRPC.js [hunk 5]
```diff
@@ -287,6 +287,33 @@ FFW.UI = FFW.RPCObserver.create(
...
+          {
+            var targetID = request.params.cancelID;
+            var typeID = request.params.functionID;
+
+            if (typeID === 10 && SDL.InteractionChoicesView.active
+               && (targetID === undefined || targetID === SDL.InteractionChoicesView.cancelID)) {
+              SDL.InteractionChoicesView.deactivate('ABORTED');
+            } else if (typeID === 12 && SDL.AlertPopUp.active
+               && (targetID === undefined || targetID === SDL.AlertPopUp.cancelID)) {
+              SDL.AlertPopUp.deactivate();
```
Comment (jacobkeeler) [inline ffw/UIRPC.js]:
```suggestion
              SDL.AlertPopUp.deactivate("ABORTED");
```
These are terribly inconsistent

**Pull request merged**